### ISSUE TITLE
Please check the 'jti' claim.

### ISSUE THREAD

**Issue by garyforsterio** [box/box-node-sdk#138]
Hello,

Using `getAppAuthClient` I'm having an issue with tokens using this SDK. I'm getting some 400s back after successive calls to box with the error body
```javascript
{ error: 'invalid_grant',
     error_description: 'Please check the \'jti\' claim.' },
```

Having read previous issues I thought this would be due to my computers clock being out of sync but using https://time.is/ I see my clock is accurately sync to some minute amount. To further investigate I console logged the jti claim. Here it is

```javascript
{ exp: 1496157257, box_sub_type: 'user' }
```
Here is the response:

```javascript
res { statusCode: 400,
  body: 
   { error: 'invalid_grant',
     error_description: 'Please check the \'jti\' claim.' },
  headers: 
   { server: 'ATS',
     date: 'Tue, 30 May 2017 15:13:20 GMT',
     'content-type': 'application/json',
     'content-length': '77',
     'cache-control': 'no-store',
     'strict-transport-security': 'max-age=31536000; includeSubDomains',
     vary: 'Accept-Encoding',
     'set-cookie': 
      [ 'box_visitor_id=592d8c0fed6377.84891543; expires=Wed, 30-May-2018 15:13:19 GMT; Max-Age=31536000; path=/; domain=.box.com; secure',
        'bv=OPS-41777; expires=Tue, 06-Jun-2017 15:13:19 GMT; Max-Age=604800; path=/; domain=.app.box.com; secure',
        'cn=28; expires=Wed, 30-May-2018 15:13:19 GMT; Max-Age=31536000; path=/; domain=.app.box.com; secure',
        'site_preference=desktop; path=/; domain=.box.com; secure' ],
     age: '2',
     connection: 'keep-alive' },
  request: 
   { uri: 
      Url {
        protocol: 'https:',
        slashes: true,
        auth: null,
        host: 'api.box.com',
        port: 443,
        hostname: 'api.box.com',
        hash: null,
        search: null,
        query: null,
        pathname: '/oauth2/token',
        path: '/oauth2/token',
        href: 'https://api.box.com/oauth2/token' },
     method: 'POST',
     headers: 
      { 'User-Agent': 'Box Node.js SDK v1.4.2',
        'content-type': 'application/x-www-form-urlencoded',
        accept: 'application/json',
        'content-length': 816 } } }
```

Comparing the exp claim to to the header date in the response I see that there is a difference of 56 seconds which is below 60, the maximum so I'm confused as to what the issue is.

Any ideas?

Cheers,
G

**Comment by mattwiller** [comment]
@garyforsterio The `jti` claim is a unique identifier generated by the SDK for the JWT; it shouldn't be related to the expiration time issues other people have run into.  For example, a JWT that I just generated with the SDK has the following claim: `"jti": "3ec3b35f-6b5b-46e2-a59b-8f19627016fa"`.

First, what version of the SDK are you using?  I don't think we've made any changes that should affect the JWT grant that way, but it never hurts to check.

Second, If you `console.log()` the JWT assertion at `lib/token-manager.js:330` and paste it into https://jwt.io you should be able to inspect exactly what is in the `jti` claim.  Could you try that and let me know if it looks different from the example I have above?

**Comment by garyforsterio** [comment]
Hey Matt,

Sorry about the slow response to this and thanks for getting back to me. I'm still having the same issue, getting 400's back with an error description `"Please check the 'jti' claim."`. I've decoded the JWT's associated with the http requests and the JTI claim seems to be fine ("jti": "dbf1d336-5cb6-43a1-b26a-e8a29134add6").

We have multiple servers up and running using the same box credentials to gain authorization. Do you think this could be causing the issue?

Cheers,
Gary

**Comment by mattwiller** [comment]
@garyforsterio Running multiple servers with the same Box credentials definitely should not be causing an issue.  If you don't mind, I'd like to do some debugging on the Box side to investigate what's happening on our end.  If you could send the following information to the Box email address in my Github profile, I can look into the issue further:

1. Your API key/client ID (NOT your client secret)
1. The dates/times you've seen this error occur in the last week or so
1. The version of the SDK you're using (is it still 1.4.2?)

**Comment by mattwiller** [comment]
After debugging with @garyforsterio over email, I wanted to circle back to this thread with the primary findings.

This issue is caused when an app is hitting Box API rate limits, and a request for JWT auth tokens fails with a 429 (rate limited) error.  The SDK retries the request exactly as it was first sent, resulting in the API rejecting the `jti` claim of the JWT because the API has already seen it (the `jti` claim is supposed to be unique).

It appears that one possible cause of this rate limiting issue is that the Event Stream object is making too many API calls and hitting the rate limits, so I'm working on #165 to eliminate that proximal cause of this issue.  The root cause can be addressed either in the Box API or in this SDK in the future, but is a more involved fix either way and may take more time to implement.

**Comment by mattwiller** [comment]
@garyforsterio To close the loop on this, we just published v1.8.0 of the SDK, which includes fixes to the Event Stream that should help with the problem you were seeing here.
